ytypes/string_type.go

```diff
@@ -26,6 +26,35 @@ import (
 
 // Refer to: https://tools.ietf.org/html/rfc6020#section-9.4.
 
+var (
+	POSIXregexCache = map[string]*regexp.Regexp{}
```

Comment:
Is it right to have this as a global -- I think that this might bring us some challenges if someone were to do something like:

```golang
d := &oc.Device{}
e := &oc.Device{}

go func() {
   _ = d.Validate()
}()

go func() {
    _ = e.Validate()
}()
```

In this case, you have two goroutines that are potentially changing the global map at the same time.

Should we rather have this such that there is some validation context that is passed between different validations within the same call? Or, alternatively, have it such that we create the cache as part of the generated package, and protect it with a mutex?

Reply:
Thinking about this - it's a bit harder to take it from the generated code whilst keeping the same function signatures that we have today - but we could add a `ValidationOpt` that covers this. So perhaps I'd suggest that this could be something that is created in the first container validation and then handed off -- this would mean that we perhaps don't need the same concurrency protection because there'll be a single context within which we're doing a validation - given that we do not parallelise the validation today.

Reply:
On a second glance, this is why the race tests are failing AFAICS.

Reply:
So I see that you're concerned about memory usage by the cache, and that we don't want to have the program store compiled version of the regexes for another part of the schema when we might now be working with a different part?

My initial thought was that since the generated code already unzips the schema, that a proportional increase in background memory use would not likely cause a problem.

So, could we just keep the global caches in `ytypes` and just introduce a mutex to solve the race condition?

----

I understand you to mean that we introduce a new type `StringValidationOpt` that would store the caches, and it's only an argument to `ValidateStringRestrictions`. The other validation functions always invokes this option when calling `ValidateStringRestrictions`? I'm fine with this, although I would prefer to not introduce this flag. I'm not quite convinced having a global cache would be so bad.

If instead, we introduce a `ValidationOpt` behaviour that's not default for all other validation functions, then it's conceivable the user to experience the following:
1. Don't use it initially since it takes some reading to become aware of it, and then some thought to figure out whether it applies to their situation since most want to minimize their code complexity.
2. Notice a performance problem (at which time it's already too late).
3. Debug their performance problem.
4. Look at the docs to find that this applies to their situation, and then uses it.

----

To make this easier for consuming packages like `ygot`, we can optionally move the regex compilation and caching down one level to `goyang` by introducing a new exported slice variable for each pattern type:
```
type YangType struct {
...
	Pattern          []string
	POSIXPattern     []string
        CompiledPatterns []*regexp.Regexp
        CompiledPOSIXPatterns []*regexp.Regexp
}
```
+ backwards compatible change without needing a new `ValidationOpt` of any kind.
+ Caching is preserved across `Validate` calls.
- Regexes always get compiled, so will incur a performance cost when constructing the AST. The user may then not ever use them. <- my guess is this is not a practical problem since the user almost always wants the patterns to be compiled as is.

However, I wonder, even though Go is backwards compatible, if the `*regexp.Regexp` were serialized and then deserialized by a different Go version, could it break?

I don't find this solution to be practically much different than using a global cache, however.

Reply:
I'm not concerned about memory usage, I'm concerned about whether the map is safe across different goroutines, because we currently don't have anything that is outside of the context of a single validation call today.  We could introduce mutex-protected global caches in the `ytypes` package for sure - but what I was thinking is whether we want to do this in the context of a particular validation such that we don't end up with locking in between different validation calls -- how we'd do this would be something that we can figure out if we agree what the scope should be. (I understand the downsides with a `ValidationOpt` - I was just discussing the different ways that we might change away from the global map.)

On the approach of moving the compilation into goyang -- I'm not clear that this is the right thing to do. It introduces a new dependency for all consumers of this library - whereas only a few might want to do this. However, one thing we could do is move to something like this when parsing the regexps.

Reply:
Ok I see. Currently, there isn't any practical difference between using a global cache vs. keeping the context within a single `Validate` call. I missed this and so assumed you meant space, apologies.

Locking and unlocking might introduce performance issues, since reading also requires locking, so using a global cache might itself cause a performance problem. I'll work on a benchmark test to help resolve this question.

Reply:
I think that using a global cache has some upsides though, right? Since you don't need to call `Compile`/`CompilePOSIX` more than exactly once per regexp in the model set that is compressed in the generated package. So there's some savings in CPU time during building the map (which I think we're saying is the slow thing today).

Building the cache once per `Validate` is 'safer' in terms of not having potential lock contention if this is done in parallel, but it then means we spend CPU time compiling regexps multiple times - the number of times that `Validate` is called.

If we consider a use case of:

 - Generate 1,000 `oc.Interface{}`, and validate each of them in turn.
 - Merge these interfaces into an `oc.Device{}` and then validate this.

which is something that we know we have - then it seems like we're better to have a global cache, since this will mean just one compile for each regexp vs. 1001 that would be needed in the following case.

Further to this, I was comparing this to what we did for the leafref case - and concluded that they are different (I'd originally considered them quite equivalent). In the leafref case we are caching results based on a lookup _in a particular data tree_, but in this case, we're caching results based on _a particular schema tree_, so it makes sense that it's generally OK to store this as a global for the `ytypes` package.

Reply:
FYI. The ZTC use case is:

1. Config gen of multiple entities are run in parallel.
2. For each config gen, we only validate the final OC device.

Therefore, from ZTC perspective, I would prefer per-Validate cache. If we implement global cache with mutex, we essentially serialize the validation calls, which could affect the performance of metroshard config gen (as metroshard is expanded into EK_NETWORK + mutiple EK_DEVICE in ZTC). On the other hand, since ZTC has stopped validating each OC piece (another performance improvement one year ago), the downside of per-Validate cache does not matter that much.

Reply:
PTAL, I've added the mutex and a benchmark test. Also see comment below.